src/rewards/mod.rs

```diff
@@ -143,102 +162,194 @@ impl<'a> RewardsMonitor<'a> {
+                // Write to cache in chunks of 100 at a time.
+                self.cache.add_epoch_data(current_epoch, queried.clone())?;
+            }
+
+            // Extend accounts by voter
+            apys.extend(
+                queried
                     .into_iter()
-                    .map(|((pk, _), a)| (pk, a))
-                    .collect::<HashMap<_, _>>();
+                    .map(|(_, (voter, percent))| ((voter, current_epoch), percent)),
+            );
+        }
 
-                // Write to cache in chunks of 100 at a time.
-                // FIXME: This writes all account state into database, which is inefficient. Refactor so that only APY data is written and `accounts` only stores APY data and drop account information here.
-                self.cache.add_epoch_data(current_epoch, insert)?;
+        // A mapping of pubkeys to APYs in the preceding `MAX_EPOCH_LOOKBACK` epochs.
+        let mut voter_epoch_apys: HashMap<Pubkey, BTreeMap<Epoch, f64>> = HashMap::new();
+        // Fill in the epoch APYs of voters.
+        for ((voter, epoch), apy) in apys {
+            voter_epoch_apys
+                .entry(*voter)
+                .and_modify(|epoch_apys| {
+                    epoch_apys.insert(*epoch, *apy);
+                })
+                .or_insert_with(|| std::iter::once((*epoch, *apy)).collect());
+        }
+
+        // TODO: Update this part according to changes to `epoch_duration_days`. A local map could
+        // become redundant if the struct caches it in a field, for example.
+        let epoch_durations = (current_epoch - MAX_EPOCH_LOOKBACK + 1..=current_epoch)
+            .map(|epoch| Ok((epoch, self.epoch_duration_days(epoch, current_epoch_info)?)))
+            .collect::<anyhow::Result<BTreeMap<_, _>>>()?;
+        let duration_max_epoch_lookback: f64 = epoch_durations.values().sum();
+
+        let mut voter_apys = HashMap::new();
+        for (voter, epoch_apys) in voter_epoch_apys {
+            let mut total_apy = 0.0;
+            for (epoch, duration) in &epoch_durations {
+                let apy = *epoch_apys.get(epoch).unwrap_or(&0.0);
+                total_apy += apy * duration;
             }
+            let average_apy = total_apy / duration_max_epoch_lookback;
+            let current_apy = *epoch_apys.get(&current_epoch).unwrap_or(&0.0);
+            voter_apys.insert(
+                voter,
+                VoterApy {
+                    current_apy,
+                    average_apy,
+                },
+            );
+        }
+        Ok(voter_apys)
+    }
 
-            // Extend accounts
-            accounts.extend(pka);
+    fn epoch_duration_days(&self, epoch: Epoch, epoch_info: &EpochInfo) -> anyhow::Result<f64> {
+        // TODO: Use some extrapolation magic to find out the epoch length if it's the current epoch
+        // If it's the current epoch we can't use the next epoch.
+        if epoch == epoch_info.epoch {
+            return Ok(3.0);
```

Comment:
Can you make it to be the last epoch's duration instead?

Reply:
Since the rewards for epoch `N-1` are in epoch `N`, it should be up to the caller of the function to offset the epochs by one (this is already done by all existing function calls). That said, the extrapolation should be easy to implement.

Reply:
Should be fixed in d34c0a68. Looks good?